modules/text/src/text_detector_swt.cpp

```diff
@@ -0,0 +1,812 @@
+        Component compi = components[i];
+        for (unsigned int j = i+1; j < components.size(); j++) {
+            Component compj = components[j];
+            if ((compi.median / compj.median <= 2.0 || compj.median / compi.median <= 2.0)
+                && (compi.width/compj.width <= 2.0 || compj.width/compi.width <= 2.0)) {
+                    float dist = (compi.cx - compj.cx) * (compi.cx - compj.cx) +
+                                    (compi.cy - compj.cy) * (compi.cy - compj.cy);
+                    float colorDist = (colorAverages[i].Red - colorAverages[j].Red) * (colorAverages[i].Red - colorAverages[j].Red) +
+                                    (colorAverages[i].Green - colorAverages[j].Green) * (colorAverages[i].Green - colorAverages[j].Green) +
+                                    (colorAverages[i].Blue - colorAverages[j].Blue) * (colorAverages[i].Blue - colorAverages[j].Blue);
+                    if (dist < 9*(float)(std::max(std::min(compi.length,compi.width),std::min(compj.length,compj.width)))
+                        *(float)(std::max(std::min(compi.length,compi.width),std::min(compj.length,compj.width))) && colorDist < 1600) {
+                            ChainedComponent chain;
+                            chain.chainIndexA = i;
+                            chain.chainIndexB = j;
+                            vector <int> componentIndices;
+                            componentIndices.push_back(i);
+                            componentIndices.push_back(j);
+                            chain.componentIndices = componentIndices;
+                            chain.chainDist = dist;
+
+                            float dx = compi.cx - compj.cx;
+                            float dy = compi.cy - compj.cy;
+                            float mod = sqrt(dx * dx + dy * dy);
+                            dx = dx / mod;
+                            dy = dy / mod;
+
+                            Direction dir;
+                            dir.x = dx;
+                            dir.y = dy;
+                            chain.dir = dir;
+                            chains.push_back(chain);
+                            count++;
+                        }
+
+            }
+        }
+    }
+
+    std::sort(chains.begin(), chains.end(), chainSortDist);
+
+    const float alignmentThreshold = CV_PI / 6;
+    int merges = 1;
+    while (merges > 0) {
+        for (unsigned int i = 0; i < chains.size(); i++) {
+            chains[i].merged = false;
+        }
+        merges = 0;
+        std::vector<ChainedComponent> chainsAfterMerging;
+        for (unsigned int i = 0; i < chains.size(); i++) {
+            for (unsigned int j = 0; j < chains.size(); j++){
+                if (i!=j && !chains[i].merged && !chains[j].merged) {
+                    if (chains[i].chainIndexA == chains[j].chainIndexA) {
+                        if (acos(chains[i].dir.x * -chains[j].dir.x + chains[i].dir.y * -chains[j].dir.y) < alignmentThreshold) {
```

Comment:
`if (acos(...expr...) < alignmentThreshold)`

if expr > 0 then:

```
const double alignmentThreshold_cos = cos(alignmentThreshold); // cos(PI/6) = 0.5 btw
...
if (expr > alignmentThreshold_cos)
```

can be used.